### DISCUSSION TITLE
Registering send event callbacks inside of actions

### DISCUSSION THREAD

**Discussion by Ghirigoro** [statelyai/xstate#3642] [Q&A] [answered]
I have a machine that manages a websocket connection. When the machine enters its `CONNECTING` state it creates and initializes a websocket within an action. The issue I'm running into is that I need to register callbacks with the websocket that will send events back to the interpreted machine. E.g:
```
// NOTE: I know `send` doesn't imperatively send events and this doesn't work ;)
socket.onopen = ()=> send("CONNECTED"); 
```
How would I register callbacks like this from within an action?

**Comment by Ghirigoro** [comment, marked as answer]
invoked services seems to be the intended way to do this. See https://github.com/statelyai/xstate/issues/434

**Reply by davidkpiano** [reply]
That's correct. Actions are fire-and-forget effects, and it seems that you don't actually want to forget the effect here.
